src/Proto3/Suite/JSONPB/Class.hs

```diff
@@ -0,0 +1,193 @@
+import           Data.Proxy
+import           Data.Text                        (Text)
+import qualified Data.Text.Lazy                   as TL
+import qualified Data.Text.Lazy.Encoding          as TL
+import           Proto3.Suite.Class               (HasDefault (def, isDefault),
+                                                   Named (nameOf))
+
+-- * Typeclass definitions
+
+-- | 'A.ToJSON' variant for JSONPB direct encoding via 'A.Encoding'
+class ToJSONPB a where
+  -- | 'A.toEncoding' variant for JSONPB encoders. Equivalent to 'A.toEncoding'
+  -- if an implementation is not provided.
+  toEncodingPB :: Options -> a -> A.Encoding
+  default toEncodingPB :: (A.ToJSON a) => Options -> a -> A.Encoding
+  toEncodingPB _ = A.toEncoding
+
+-- | 'A.FromJSON' variant for JSONPB decoding from the 'A.Value' IR
+class FromJSONPB a where
+  -- | 'A.parseJSON' variant for JSONPB decoders. Equivalent to 'A.parseJSON' if
+  -- an implementation is not provided.
+  parseJSONPB :: A.Value -> A.Parser a
+
+  default parseJSONPB :: (A.FromJSON a) => A.Value -> A.Parser a
+  parseJSONPB = A.parseJSON
+
+-- * JSONPB codec entry points
+
+-- | 'Data.Aeson.encode' variant for serializing a JSONPB value as a lazy
+-- 'LBS.ByteString'.
+encode :: ToJSONPB a => Options -> a -> LBS.ByteString
+encode opts = E.encodingToLazyByteString . toEncodingPB opts
+{-# INLINE encode #-}
+
+-- | 'Data.Aeson..eitherDecode' variant for deserializing a JSONPB value from a
+-- lazy 'LBS.ByteString'.
+eitherDecode :: FromJSONPB a => LBS.ByteString -> Either String a
+eitherDecode = eitherFormatError . A.eitherDecodeWith jsonEOF (A.iparse parseJSONPB)
+  where
+    eitherFormatError = either (Left . uncurry A.formatError) Right
+    {-# INLINE eitherFormatError #-}
+
+    -- NB: cribbed from aeson-1.1.1.0:Data.Aeson.Parser.Internal.jsonEOF, which
+    -- is not exported. It's simple, so we just inline it here. Might be worth
+    -- submitting a PR to export this.
+    jsonEOF :: Atto.Parser A.Value
+    jsonEOF = A.json <* skipSpace <* Atto.endOfInput
+      where
+        skipSpace :: Atto.Parser ()
+        skipSpace = Atto.skipWhile $ \w -> w == 0x20 || w == 0x0a || w == 0x0d || w == 0x09
+        {-# INLINE skipSpace #-}
+{-# INLINE eitherDecode #-}
+
+-- * Operator definitions
+
+-- | Construct a (field name, JSONPB-encoded value) tuple
+(.=) :: (HasDefault v, ToJSONPB v) => Text -> v -> FieldPB
+k .= v = FieldPB (isDefault v) (\opts -> E.pair k (toEncodingPB opts v))
+
+-- | 'Data.Aeson..:' variant for JSONPB; if the given key is missing from the
+-- object, or if it is present but its value is null, we produce the default
+-- protobuf value for the field type
+(.:) :: (FromJSONPB a, HasDefault a) => A.Object -> Text -> A.Parser a
+obj .: key = obj .:? key A..!= def
+  where
+    (.:?) = A.explicitParseFieldMaybe parseJSONPB
+
+-- * JSONPB rendering and parsing options
+
+data Options = Options
+  { optEmitDefaultValuedFields :: Bool
+  }
+
+defaultOptions :: Options
```

Comment:
I've found that it can be really nice (not necessary though, this is a nit) in generated documentation to see what the default values are.

It creates a bigger maintenance burden and is likely not worth doing since `Options` here is such a small record, but I have found myself appreciating `defaultOptions`-style producers that document what the default values are without forcing me to go look at the source (I know, I'm lazy).